components/notebook-controller/pkg/culler/culler.go

```diff
@@ -0,0 +1,194 @@
+package culler
+
+import (
+	"encoding/json"
+	"fmt"
+	"net/http"
+	"os"
+	"strconv"
+	"time"
+
+	metav1 "k8s.io/apimachinery/pkg/apis/meta/v1"
+	logf "sigs.k8s.io/controller-runtime/pkg/runtime/log"
+)
+
+var log = logf.Log.WithName("culler")
+var client = &http.Client{
+	Timeout: time.Second * 10,
+}
+
+// The constants with name 'DEFAULT_{ENV_Var}' are the default values to be
+// used, if the respective ENV vars are not present.
+// All the time numbers correspond to minutes.
+const DEFAULT_IDLE_TIME = "1440" // One day
+const DEFAULT_CULLING_CHECK_PERIOD = "1"
+const DEFAULT_ENABLE_CULLING = "false"
+const STOP_ANNOTATION = "kubeflow-resource-stopped"
+
+type NotebookStatus struct {
+	Started      string `json:"started"`
+	LastActivity string `json:"last_activity"`
+	Connections  int    `json:"connections"`
+	Kernels      int    `json:"kernels"`
+}
+
+// Some Utility Fuctions
+func getEnvDef(variable string, default_val string) string {
+	envVar := os.Getenv(variable)
+	if len(envVar) == 0 {
+		return default_val
+	}
+	return envVar
+}
+
+// Time / Frequency Utility functions
+func createTimestamp() string {
+	now := time.Now()
+	return now.Format(time.RFC3339)
+}
+
+func GetRequeueTime() time.Duration {
+	// The frequency in which we check if the Pod needs culling
+	// Uses ENV var: CULLING_CHECK_PERIOD
+	culling_period := getEnvDef(
+		"CULLING_CHECK_PERIOD", DEFAULT_CULLING_CHECK_PERIOD)
+	culling_period_m, err := strconv.Atoi(culling_period)
+	if err != nil {
+		log.Info(fmt.Sprintf(
+			"Culling Period should be Int. Got '%s'. Using default value.",
+			culling_period))
+		culling_period_m, _ = strconv.Atoi(DEFAULT_CULLING_CHECK_PERIOD)
+	}
+
+	return time.Duration(culling_period_m) * time.Minute
+}
+
+func getMaxIdleTime() time.Duration {
+	idle_time := getEnvDef("IDLE_TIME", DEFAULT_IDLE_TIME)
+	idle_time_m, err := strconv.Atoi(idle_time)
+	if err != nil {
+		log.Info(fmt.Sprintf(
+			"IDLE_TIME should be Int. Got %s instead. Using default value.",
+			idle_time))
+		idle_time_m, _ = strconv.Atoi(DEFAULT_IDLE_TIME)
+	}
+
+	return time.Minute * time.Duration(idle_time_m)
+}
+
+// Stop Annotation handling functions
+func SetStopAnnotation(meta *metav1.ObjectMeta) {
+	if meta == nil {
+		log.Info("Error: Metadata is Nil. Can't set Annotations")
+		return
+	}
+
+	if meta.GetAnnotations() != nil {
+		annotations := meta.GetAnnotations()
```

Comment:
So this gets a copy of the map, and you need to call set again?
Then do you need to call set in the RemoveStopAnnotation below?

Reply:
Simplified it to edit the `meta.Annotations` directly and not use a Get and then a Set method.
I think it's ok in the `RemoveStopAnnotation`.